tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -25,25 +29,190 @@ pub struct CudaVec<T: Numeric> {
 }
 
 impl<T: Numeric> CudaVec<T> {
+    pub fn new(len: usize, stream: &CudaStream) -> Self {
+        let vec = unsafe { Self::new_async(len, stream) };
+        stream.synchronize();
+        vec
+    }
     /// # Safety
     ///
-    /// - `ptr` __must__ be a valid device pointer to an array of `len` elements of type `T`
-    pub unsafe fn new(ptr: CudaPtr, len: usize, device: CudaDevice) -> Self {
+    /// - `stream` __must__ be synchronized to guarantee computation has finished
+    pub unsafe fn new_async(len: usize, stream: &CudaStream) -> Self {
+        let size = len as u64 * std::mem::size_of::<T>() as u64;
+        let ptr = CudaPtr {
+            ptr: cuda_malloc_async(size, stream.as_c_ptr()),
```

Comment:
Here, if `len` is `0`, size will be 0

Is it ok for cuda_malloc to allocate 0 bytes ? I've seen on the internet that is returns a null ptr in that case which is ok, can you confirm ?

Reply:
Yes it's ok :+1: